keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,294 @@
+}
+
+var alphaMetricDefinition = kubemetrics.CounterOpts{
+    Name: "some_alpha_metric",
+    Help: "some description",
+    StabilityLevel: kubemetrics.ALPHA, // this is also a custom metadata field
+	DeprecatedVersion: "1.15", // this can optionally be included on alpha metrics, although there is no change to contractual stability guarantees
+}
+```
+
+### Metric Instantiation Phase
+
+Currently, the metric instantiation phase looks like this:
+
+```go
+var someCounterVecMetric = prometheus.NewCounterVec(
+    someMetricDefinition,
+    []string{"some-label", "other-label"},
+}
+```
+
+Wrapping the prometheus constructors would allow us to take, as inputs, the modified metric definitions defined above, returning a custom kubernetes metric object which contains the metric which would have been instantiated as well as the custom metadata:
+
+```go
+var deprecatedMetric = kubemetrics.NewCounterVec( // this is a wrapped initializer, which takes in our custom metric definitions
+    deprecatedMetricDefinition, // this is our custom wrapped metric definition from above
+    []string{"some-label", "other-label"},
+}
+var alphaMetric = kubemetrics.NewCounterVec{
+    alphaMetricDefinition, // this is also our custom wrapped metric definition from above
+    []string{"some-label", "other-label"},
+}
+```
+
+### Metric Enrollment Phase
+
+Currently, metric enrollment involves calls to a prometheus function which enrolls the metric in a global registry, like so:
+```go
+prometheus.MustRegister(someCounterVecMetric)
+```
+
+Wrapping a prometheus registry with a kubernetes specific one, would allow us to take our custom metrics from our instantiation phase and execute custom logic based on our custom metadata. Our custom registry would hold a reference to a prometheus registry and defer metric enrollment unless preconditions were met:
+
+```go
+import version "k8s.io/apimachinery/pkg/version"
+
+type Registry struct {
+    promregistry *prometheus.Registry
+    KubeVersion version.Info
+}
+
+// inject custom registration behavior into our registry wrapper
+func (r *Registry) MustRegister(metric kubemetrics.Metric) {
+    // pretend we have a version comparison utility library
+    if metricutils.compare(metric.DeprecatedVersion).isLessThan(r.KubeVersion) {
+        // check if binary has deprecated metrics enabled otherwise
+        // no-op registration
+        return
+    } else if metricutils.compare(metric.DeprecatedVersion).isEqual(r.KubeVersion) {
+        // append deprecated text to description
+        // emit warning in logs
+        // continue to actual registration
+    }
+    // append alpha text to metric description if metric.isAlpha
+    // fallback to original prometheus behavior
+    r.promregistry.MustRegister(metric.realMetric)
+}
+
+```
+
+Which we would invoke, like so:
+```go
+kubemetrics.MustRegister(deprecatedMetric)
+kubemetrics.MustRegister(alphaMetric)
+```
+
+## Stability Classes
```

Comment:
It seems to me that graduating metrics should have a defined process, just like API reviews are part of graduating an API. I feel at least defining that this must happen, and maybe a combination of the existing api-review commitee and sig-instrumentation have to give consent for something to graduate.

Reply:
Yes, that makes sense. I'll dig a bit more into the existing processes and explore this a bit.

Reply:
Added a recommendation to require api-review for graduating/deprecating metrics.